src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java

```diff
@@ -0,0 +1,608 @@
+        }
+
+        final var result = createResultFromBuildResult(report, (Participation) participation);
+        result.setRatedIfNotExceeded(report.getRunDate(), submission);
+        result.setSubmission(submission);
+
+        return Optional.empty();
+    }
+
+    private void addFeedbackToResult(Result result, TestResults report) {
+        // No feedback for build errors
+        if (report.getResults() == null || report.getResults().isEmpty()) {
+            result.setHasFeedback(false);
+            return;
+        }
+
+        final var feedbacks = report.getResults().stream().flatMap(testsuite -> testsuite.getTestCases().stream()).map(testCase -> {
+            final var feedback = new Feedback();
+            feedback.setPositive(testCase.getErrors() == null && testCase.getFailures() == null);
+            feedback.setText(testCase.getName());
+            String errorMessage = null;
+            // If we have errors or failures, they will always be of length == 1 since JUnit (and the format itself)
+            // should generally only report he first failure in a test case
+            if (testCase.getErrors() != null) {
+                errorMessage = testCase.getErrors().get(0).getMessage();
+            }
+            else if (testCase.getFailures() != null) {
+                errorMessage = testCase.getFailures().get(0).getMessage();
+            }
+            // The assertion message can be longer than the allowed char limit, so we shorten it here if needed.
+            if (errorMessage != null && errorMessage.length() > FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS) {
+                errorMessage = errorMessage.substring(0, FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS);
+            }
+            feedback.setDetailText(errorMessage);
+
+            return feedback;
+        }).collect(Collectors.toList());
+
+        result.setHasFeedback(true);
+        result.addFeedbacks(feedbacks);
+    }
+
+    private TestResults fetchLatestBuildResultFromJenkins(ProgrammingExerciseParticipation participation) {
+        final var projectKey = participation.getProgrammingExercise().getProjectKey();
+        final var planKey = participation.getBuildPlanId();
+        final var url = Endpoint.TEST_RESULTS.buildEndpoint(JENKINS_SERVER_URL.toString(), projectKey, planKey).build(true);
+
+        return restTemplate.getForObject(url.toUri(), TestResults.class);
+    }
+
+    @Override
+    public List<Feedback> getLatestBuildResultDetails(Result result) {
+        // TODO since this is unused as of now
+        throw new NotImplementedException("Jenkins service does not support fetching the latest feedback for a result");
+    }
+
+    @Override
+    public List<BuildLogEntry> getLatestBuildLogs(String projectKey, String buildPlanId) {
+        try {
+            final var build = job(projectKey, buildPlanId).getLastBuild();
+            final var logHtml = Jsoup.parse(build.details().getConsoleOutputHtml()).body();
```

Comment:
I am just curious here if it is the easiest way to get the build log from html? Is the output text not better for this?

Reply:
The problem is that the raw output only contains the text nodes, so we don't get the timestamps. Only the HTML contains those.